scipy/stats/_unuran/unuran_wrapper.pyx.templ

```diff
@@ -0,0 +1,1122 @@
+            msg = self._messages.get()
+            if msg:
+                raise UNURANError(msg)
+        finally:
+            _lock.release()
+            release_unuran_callback(&callback)
+
+    def ppf_hat(self, u):
+        """
+        Evaluate the inverse of the CDF of the hat distribution at `u`.
+
+        This call does not work for variant IA (immediate acceptance).
+        In this case the hat CDF is evaluated as if variant PS is used.
+
+        Parameters
+        ----------
+        u : array_like
+            An array of percentiles
+
+        Returns
+        -------
+        ppf_hat : array_like
+            Array of quantiles corresponding to the given percentiles.
+
+        Examples
+        --------
+        >>> from scipy.stats import TransformedDensityRejection
+        >>> from scipy.stats import norm
+        >>> from math import exp
+        >>> 
+        >>> class MyDist:
+        ...     def pdf(self, x):
+        ...         return exp(-0.5 * x**2)
+        ...     def dpdf(self, x):
+        ...         return -x * exp(-0.5 * x**2)
+        ... 
+        >>> dist = MyDist()
+        >>> rng = TransformedDensityRejection(dist)
+        >>> 
+        >>> rng.ppf_hat(0.5)
+        -0.00018050266342393984
+        >>> norm.ppf(0.5)
+        0.0
+        >>> u = np.linspace(0, 1, num=1000)
+        >>> ppf_hat = rng.ppf_hat(u)
+        """
+        u = np.asarray(u, dtype='d')
+        oshape = u.shape
+        u = u.ravel()
+        # UNU.RAN fills in ends of the support when u < 0 or u > 1 while
+        # SciPy fills in nans. Prefer SciPy behaviour.
+        cond0 = 0 <= u
+        cond1 = u <= 1
+        cond2 = cond0 & cond1
+        goodu = argsreduce(cond2, u)[0]
+        out = np.empty_like(u)
+        cdef double[::1] u_view = np.ascontiguousarray(goodu)
+        cdef double[::1] goodout = np.empty_like(u_view)
+        if cond2.any():
+            self._ppf_hat(&u_view[0], &goodout[0], len(goodu))
+        np.place(out, cond2, goodout)
+        np.place(out, ~cond2, np.nan)
+        return np.asarray(out).reshape(oshape)[()]
+
+
+cdef class DiscreteAliasUrn(Method):
+    r"""
+    Discrete Alias-Urn Method.
+
+    This method is used to sample from univariate discrete distributions with
+    a finite domain. It uses the probability vector of size :math:`N` or a
+    probability mass function with a finite support to generator random
+    numbers from the distribution.
+
+    Parameters
+    ----------
+    pv : array_like, optional
+        Probability vector (PV) of the distribution. If PV isn't available,
+        PMF of the distribution is expected.
+    dist : object, optional
+        An instance of a class with a 'pmf' method. Only required if PV isn't
+        available. The signature of the PMF is expected to be:
+        ``def pmf(k: int, *params) -> float``. i.e it should accecpt a Python
+        integer and return a Python float.
```

Comment:
Do we need both options? It seems that the user has `pv`, then they can create an `rv_discrete` out of it. 

If `pv` is provided and `domain` is ignored, is it assumed that the inputs `x` to which the `pv`s correspond are `np.arange(len(pv))`? Why not use `domain`, e.g. `np.arange(domain[0], domain[1]+1)`?

Reply:
> Do we need both options? It seems that the user has `pv`, then they can create an `rv_discrete` out of it we were only to accept `dist`. (And if the user were to provide an object like an `rv_discrete`, a `domain` input wouldn't really be needed.)

This is a difficult decision. I figured this boils down to two different types of UI: 

1. A `dist` objects holds the distribution methods (e.g. `pmf`, `pdf`, `cdf`, etc) and its attributes (e.g. `domain`, `pv`, `mode`, etc)
2. A `dist` object just holds the distribution methods.

Originally, I just passed all the methods and attributes separately as parameters. But to be more consistent with `NumericalInverseHermite` and pack all the methods in one object, we added a `dist` parameter. The natural choice here was to go for point 2 because we just wanted to use the `dist` object as a way to batch all the required methods in one object. This way one can pass a `rv_discrete` or `rv_continuous` object if the required methods are present without subclassing it and adding more attributes (like `pv`, `domain`, `mode`, `pdfarea`, etc).

On the other hand, now that I think of it, it might be better to unify all the methods and attributes in one single `dist` object with the only ramification being inconsistency with the `rv_discrete` and `rv_continuous` classes i.e. users will need to create their distribution object from scratch or subclass existing SciPy distributions to add required attributes and methods on top of it (or, as a workaround, add attributes directly without subclassing). 

Let me know which option sounds better.

<hr>

> If `pv` is provided and `domain` is ignored, is it assumed that the inputs `x` to which the `pv`s correspond are `np.arange(len(pv))`?

Yes, exactly.

> Why not use `domain`, e.g. `np.arange(domain[0], domain[1]+1)`?

I left this choice up to UNU.RAN. By default, UNU.RAN chooses (0, len(pv)) as the domain on which the distribution is defined. I thought it made sense and didn't change it. Though, changing it should not be really difficult. If you think (1, len(pv) + 1) makes a better choice for the domain, I can change it that way.

Reply:
Nicholas, Tirth and I talked about this in our call today. We thought it is better to avoid adding more attributes like `mode` and `domain` to `dist`:

writing `TransformedDensityRejection(..., domain=(1,2))` is more transparent than having a domain attribute in `dist` that is not directly visible. in general, domain can be used to truncate a distribution, i.e., it is not part of the underlying distribution.

`pv` might be an edge case where one could argue that the user should create a dist object from it first and then pass it to TDR instead of `pv`.

Reply:
> We thought it is better to avoid adding more attributes like...

OK, I think this got broadened a bit from my original comment, though. I'm just talking about `pv` and `dist` seeming redundant. Maybe a compromise would be accepting only `dist` and checking for an attribute `pmf`. If it's not there, it had better be an array-like so it can be treated like a PV. The possibility of calling the function with both - especially as positional arguments - just seems funny to me.

Reply:
> If you don't want to force users to create an `rv_discrete` (or custom object with `pmf` method) out of a PV before passing it in, maybe a compromise would be to accept only one argument `dist` and deciding which it is.

I agree. I remember, at some point, I did that with the `pv` and `pmf` method and called the parameter `pv_or_pmf`. We can do something like that here -> `pv_or_dist`.

> The possibility of calling the function with both `pv` and `dist` - especially as positional arguments with `pv` first (instead of `dist`, which is the first argument of the other random sampling classes) - just disagrees with me.

True. I will change this. Thanks for the suggestion.

Reply:
I think just `dist` would be fine. The probability vector is about as complete of a representation of a distribution as an object with a `pmf` method, so I don't think `dist` is an unreasonable name.

Reply:
done.

Reply:
This looks much better. There are a few things I'm still thinking about, though. You wrote:

> in general, `domain` can be used to truncate a distribution, i.e., it is not part of the underlying distribution.

Which means that if `dist` is an object with a `pmf` method, then `domain` is the only way of setting its domain; any `support` method (which all `rv_discrete`s have) will be ignored. However,

> When this parameter is passed in combination with a probability vector, `domain[0]` is used to relocate the distribution from `(0, len(pv)) to (domain[0], domain[0]+len(pv))` and `domain[1]` is ignored. See Notes and tutorial for a more detailed explanation.

So `domain` is _not_ used to truncate a PV's domain. 

The things that are not sitting right with me are:
1. `domain[1]` is ignored for PVs but not for distribution objects
2. the user _can't_ rely on a distribution object's `support` method

2 could be addressed by the following logic: if `domain` is not provided, and if `dist` is a distribution object, use its `support` method to determine the domain when doing input validation. 

1 is trickier. You _could_ use `domain` to truncate a PV, but I can imagine three conventions for doing that:
* `domain[0]` is used to locate the left end of the distribution; `domain[1]` is used to trim the right end.
* `domain[1]` is used to locate the right end of the distribution; `domain[0]` is used to trim the left end.
* `domain[0]` and `domain[1]` are used as indices of `dist`; they cannot be used to locate the distributon. If the user wants to change the location, they can add a constant to the random variates. 

While perhaps not as situationally convenient, the last option seems to be the most consistent to me, as it is a much closer analog of what `domain` means when `dist` is a distribution object.

Not that it's a _good_ option, but another option would be to eliminate `domain` altogether. For a PV, the user controls the domain by slicing their original PV and adding a constant to the random variates. There is some precedent for this - I imagine everyone has transformed the output of `np.random.rand` with some constant and scale at some point. For a distribution object, the user controls the domain by redefining `support`. Not saying it's a convenient option, but it is an option if we're not sure how this should work right now.

I'm uncomfortable with ignoring `support`, though. This is part of SciPy, so I think it should work in an intuitive way with SciPy's distribution objects.

Reply:
Good point. When I read your comment, I first agreed that it is weird. I just checked and noted that the current implementation mimics the behaviour in UNU.RAN:

> Set the left and right borders of the domain of the distribution. This can also be used to truncate an existing distribution. For setting the boundary to +/- infinity use INT_MIN and INT_MAX, respectively. [...]If the PV of the discrete distribution is used, than the right boudary is ignored (and internally set to left + size of PV - 1). Notice that INT_MIN and INT_MAX are interpreted as (minus/plus) infinity.

I think I now understand better why it is done like this:
- if a distribution is given, it has a built-in support and domain can be used to truncate it.
- if a pv is used, one could say that truncation is not really important: in that case, the user should just use the method with the rescaled probability vector: if `ind` are the indices of `pv` that the distributions should be restricted to, just use `pv_new = pv[ind] / sum(pv[ind])`. so instead of truncation, domain[0] is used to shift the distribution. 

I hope it makes sense. I understand that this is a bit odd, but there is a rationale behind it. If we document it properly, it is ok in my view. what do you think?

> 2 could be addressed by the following logic: if domain is not provided, and if dist is a distribution object, use its support method to determine the domain when doing input validation.

I agree (and implicitly assumed that is how it works :) @tirthasheshpatel do you see a reason against this approach?

Reply:
> I agree (and implicitly assumed that is how it works :) @tirthasheshpatel do you see a reason against this approach?

I think this is a nice idea too. We can use the `support` method when a domain is not explicitly provided.

Reply:
This comment belongs here: The implementation now supports the attribute support :)

With the updated documentation of the approach if PV is used, is that ok for you (@mdhaber)?